packages/client/src/features/tools/edge-creation/edge-creation-tool.ts

```diff
@@ -128,33 +141,74 @@ export class EdgeCreationToolMouseListener extends DragAwareMouseListener {
             if (this.currentTarget) {
                 if (!this.isSourceSelected()) {
                     this.allowedTarget = this.isAllowedSource(newCurrentTarget);
                 } else if (!this.isTargetSelected()) {
-                    this.allowedTarget = this.isAllowedTarget(newCurrentTarget);
-                }
-                if (this.allowedTarget) {
-                    const action = !this.isSourceSelected()
-                        ? cursorFeedbackAction(CursorCSS.EDGE_CREATION_SOURCE)
-                        : cursorFeedbackAction(CursorCSS.EDGE_CREATION_TARGET);
-                    return [action];
+                    // Temporarily mark the target as invalid while we check for a proper result,
+                    // so a fast-clicking user doesn't get a chance to create the edge in the meantime.
+                    this.allowedTarget = false;
+                    const actions = this.isAllowedTarget(newCurrentTarget).then(allowedCheck => {
+                        // Make sure we didn't change the target element while
+                        // checking for valid target.
+                        if (allowedCheck.targetElement === this.currentTarget) {
+                            this.allowedTarget = allowedCheck.allowed;
+                            return this.updateEdgeFeedback();
+                        }
+                        // FIXME Is there a proper no-op action? We can't return <undefined> or [] here
+                        // because of the method signature.
+                        return <Action>{
+                            kind: 'no-op'
+                        };
+                    });
+                    return [actions];
                 }
+                return [this.updateEdgeFeedback()];
             }
-            return [cursorFeedbackAction(CursorCSS.OPERATION_NOT_ALLOWED)];
+            return [this.updateEdgeFeedback()];
         }
         return [];
     }
 
+    protected updateEdgeFeedback(): Action {
+        if (this.allowedTarget) {
+            const action = !this.isSourceSelected()
+                ? cursorFeedbackAction(CursorCSS.EDGE_CREATION_SOURCE)
+                : cursorFeedbackAction(CursorCSS.EDGE_CREATION_TARGET);
+            return action;
+        }
+        return cursorFeedbackAction(CursorCSS.OPERATION_NOT_ALLOWED);
+    }
+
     protected isAllowedSource(element: SModelElement | undefined): boolean {
         return element !== undefined && isConnectable(element) && element.canConnect(this.proxyEdge, 'source');
     }
 
-    protected isAllowedTarget(element: SModelElement | undefined): boolean {
-        return element !== undefined && isConnectable(element) && element.canConnect(this.proxyEdge, 'target');
+    protected async isAllowedTarget(element: SModelElement | undefined): Promise<AllowedTargetCheck> {
+        let allowed = element !== undefined && isConnectable(element) && element.canConnect(this.proxyEdge, 'target');
```

Comment:
I think this code does not work for purley dyanmic usecases i.e. we don't pass any allowed source/target ids
AFAIK currently the `canConnect` method of elements is overwritten by our typeHintsModule (https://github.com/eclipse-glsp/glsp-client/blob/256d64ddbb4f5596d75df654a8a93502e314d092/packages/client/src/features/hints/type-hints.ts#L124)
Am I missing something here or would `allowed` always be false if dynamic target resolution is activated? (Because can connect should be false in that case)

Reply:
Indeed; the "purely dynamic" use case is not supported. The dynamic check always comes after the regular type hint check, so if the regular check fails, the edge cannot be created.